### PR TITLE
[pledge-dashboard] feat: integrate moralis to site

### PR DESCRIPTION
**Pull request by 0xAeterno**
#### Related Epic https://github.com/KlimaDAO/klimadao/issues/255

- Integrate Moralis into `site`
- Move pledge dashboards to `/pledge/[address]` page routes
- Query for a profile based on url params

### TODO
- [x] Source Moralis account details for staging and production environments
- [ ] Documentation (TBC at a later date)

## Checklist

- [x] Building the site with `npm run build-site` works without errors
- [x] Building the app with `npm run build-app` works without errors
- [x] I formatted JS and TS files with running `npm run format-all`

### CODE REVIEW

File: site/lib/moralis.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import Moralis from "moralis/node.js";
+
+export const MoralisClient = Moralis.start({
+  appId: process.env.NEXT_PUBLIC_MORALIS_APP_ID,
```
Comment (Atmosfearful) [inline site/lib/moralis.ts:4]: I've made the following secrets available under `process.env`
image

**Review by Atmosfearful — APPROVED** [review]

File: site/lib/moralis.ts (hunk 1 above)
Comment (0xAeterno) [reply site/lib/moralis.ts:4]: @Atmosfearful 
We'll need to instantiate the client with `MORALIS_APP_ID` and `MORALIS_SERVER_URL`.
Do we need the `NEXT_PUBLIC` prefix too expose it to the client?

Comment (Atmosfearful) [reply site/lib/moralis.ts:4]: Ah derp, will do that now.

Comment (Atmosfearful) [reply site/lib/moralis.ts:4]: image

**Review by Atmosfearful — APPROVED** [review]

File: site/.env.example [hunk 2]
```diff
@@ -0,0 +1,5 @@
+// NEXT_PUBLIC prefix required to expose env vars to client
+// https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables-to-the-browser
+
+NEXT_PUBLIC_MORALIS_APP_ID=
+NEXT_PUBLIC_MORALIS_SERVER_URL=
```
Comment (0xAeterno) [inline site/.env.example:5]:
```suggestion
NEXT_PUBLIC_MORALIS_SERVER_URL=
MORALIS_SECRET=
```

File: site/queries/pledge/index.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export const getPledge = (address: string) =>
+  fetch(`http://localhost:3000/api/pledge?address=${address}`, {
```
Comment (0xAeterno) [inline site/queries/pledge/index.ts:2]: use `API_BASE_URL`

File: site/pages/pledge/[address].tsx [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import { loadTranslation } from "lib/i18n";
+
+export const getStaticProps: GetStaticProps = async (ctx) => {
+  const translation = await loadTranslation(ctx.locale);
+  const { address } = ctx.params as { address: string };
+  let pledge;
+
+  try {
+    const reponse = await getPledge(address.toLowerCase());
+    const data = await reponse.json();
```
Comment (0xAeterno) [inline site/pages/pledge/[address].tsx:14]:
```suggestion
    const response = await getPledge(address.toLowerCase());
    const data = await response.json();
```

File: site/queries/pledge/index.ts (hunk 3 above)
Comment (0xAeterno) [reply site/queries/pledge/index.ts:2]: @Atmosfearful - do we have staging/dev account credentials?

File: site/pages/pledge/[address].tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import { getPledge } from "queries/pledge";
+import { loadTranslation } from "lib/i18n";
+
+export const getStaticProps: GetStaticProps = async (ctx) => {
+  const translation = await loadTranslation(ctx.locale);
+  const { address } = ctx.params as { address: string };
+  let pledge;
+
+  try {
+    const response = await getPledge(address.toLowerCase());
```
Comment (Atmosfearful) [inline site/pages/pledge/[address].tsx:13]: Better if we use `getPledgeByAddress` here instead, to share the server-side code and avoid an extra network roundtrip, as recommended in the docs:
https://nextjs.org/docs/basic-features/data-fetching/get-static-props#write-server-side-code-directly

File: site/pages/api/pledge/index.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { NextApiRequest, NextApiResponse } from "next/types";
+import { getPledgeByAddress, pledgeResolver } from "lib/moralis";
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse
```
Comment (Atmosfearful) [inline site/pages/api/pledge/index.ts:6]: let's make sure we type the expected response data and request body. We can share the same types on the client too.

Here is some example boilerplate to make this easier, at least for inferring the correct return type:
```ts
const data = await fetchFromEndpoint<PledgeResponse>("/api/pledge", {
    body: JSON.stringify(body),
});

export const fetchFromEndpoint = async <T = any>(
  url: Endpoint,
  init?: RequestInit | undefined
): Promise<T> => {
  try {
    const res = await fetch(`${BASE_URL}${url}`, init);
    if (!res.ok) {
      const json = await res.json();
      throw json;
    }
    const json: T = await res.json();
    return json;
  } catch (e) {
    console.error(e);
    throw e;
  }
};
```

File: site/queries/pledge/index.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { urls } from "@klimadao/lib/constants";
+import { IS_LOCAL_DEVELOPMENT } from "lib/constants";
+
+const API_BASE_URL = IS_LOCAL_DEVELOPMENT ? "http://localhost:3000" : urls.home;
```
Comment (Atmosfearful) [inline site/queries/pledge/index.ts:4]: Can we export this from site/lib/constants too? Will likely be reused

File: site/queries/pledge/index.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+export const getPledge = (address: string) =>
```
Comment (Atmosfearful) [inline site/queries/pledge/index.ts:6]: Needs return types, it's probably inferring `any`. See my boilerplate example above

File: site/pages/api/pledge/index.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  switch (req.method) {
+    case "GET":
+      try {
+        const data = await getPledgeByAddress(address);
+
+        if (!data) {
+          res.status(404).json({ message: "Not found" });
+        }
+
+        const pledge = JSON.parse(JSON.stringify(data));
```
Comment (Atmosfearful) [inline site/pages/api/pledge/index.ts:19]: why stringify and parse again? Cant you just pass as-is?

File: site/pages/pledge/[address].tsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    const data = await response.json();
+    pledge = data.pledge;
+
+    if (!data.pledge) {
+      pledge = null;
+    }
+  } catch (e) {
+    console.log(e);
```
Comment (Atmosfearful) [inline site/pages/pledge/[address].tsx:21]: pledge = null

**Review by Atmosfearful — APPROVED** [review]
I can merge to keep things moving but please address these comments in follow-up PRs :)

**Pull request merged**